source/byte_buf.c

```diff
@@ -0,0 +1,113 @@
+/*
+* Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include <aws/common/byte_buf.h>
+#include <assert.h>
+#include <stdarg.h>
+
+#ifdef _MSC_VER
+/* disables warning non const declared initializers for Microsoft compilers */
+#pragma warning(disable:4204)
+#pragma warning(disable:4706)
+#endif
+
+int aws_string_split_on_char_n(struct aws_byte_buf *input_str, char split_on,
+    struct aws_array_list *output, size_t n) {
+    assert(input_str);
+    assert(output);
+    assert(output->item_size >= sizeof(struct aws_byte_cursor));
+
+    size_t max_splits = n > 0 ? n : SIZE_MAX;
+    size_t last_offset = 0, split_count = 0;
+    uint8_t *new_location = NULL;
+
+    struct aws_byte_cursor current_pos = aws_byte_cursor_from_buf(input_str);
+
+    while (split_count < max_splits &&
```

Comment:
I am refactoring the split function in CPP SDK, just checking out here.

You may want `split_count < max_splits - 1` here, or you'd better refactoring what `N `means int the comments.

If you don't hit the end when the while loop ended, the `split_count` will be `max_splits`, then you add the remaining to array. The results will have `n+1` splits(This is aligned with your test below, but the comments says, N is the max number of splits)